### DISCUSSION TITLE
Cleaner Expression body definition for simple set  accessors.

### DISCUSSION THREAD

**Discussion by Zxynine** [dotnet/csharplang#6018] [Language Ideas] [answered]
Im not sure if this is the correct place to put suggestions, I didnt want to make an issue as its more of just a request and syntactic sugar, my suggestion is to make the set accesor slightly more implicit for simple  single variable setting so it looks similar to the get accesor while using expression body definitions. 

Heres an example of the current  form factor of a basic property:
```
		public bool IsDrawerMode {
			get => isDrawerMode;
			set => isDrawerMode = value;
		}
```

In cases like this, where the property does nothing except wrap the private variable in a sense, it would be better to have the get and set look the same because as is it is sort of an eye sore with many of these in one class. Heres an example of that:
```
		public bool IsDrawerMode {
			get => isDrawerMode;
			set => isDrawerMode;
		}
```

In this case, just as the compiler recognises that you want to return the value in the getter, it would also recognise (by there only being the variable) that you wish to set the variable to the value. If you add anything to it, then the compiler will just treat it as a normally.  I think this is a bit more intuitive.

EDIT: 

I realised I provided too simple of an example that can just be implemented using {get; set;}, Here is another that should showcase why i made this suggestion:
```
		public colour StylerColour{
			get => this.styler.color;
			set => this.styler.color = value
		}
```
Would become
```
		public colour StylerColour{
			get => this.styler.color;
			set => this.styler.color;
		}
```

And here is a bit more explanation as to why I feel like this would be important:

 I feel that having it being called a setter implies clearly that you are setting the value without the need of a proper asignment in such simple cases. Just like for the getter you do not specify return. I feel like there needs to be consistency in that and we should either be able to omit "= value" in setters or should be required to include "return " in getters.

**Comment by CyrusNajmabadi** [comment]
You can just do `public bool IsDrawerMode { get; set; }` if you would like these semantics with a simple syntax.  Thanks!

**Reply by Zxynine** [reply]
Not if the variable you are wrapping is implemented somewhere else. It was my fault for including such a simple example, but what about in instances like this :

```
public colour StylerColour {
        get => this.styler.color;
        set => this.styler.color;
}
```

**Reply by HaloFour** [reply]
IMO it's much cleaner for the expression within the `set` accessor to actually be a proper assignment.  That actually describes what the accessor is doing.

If the language were to do anything here I'd rather a bidirectional delegation operator, e.g.

```
public Color StylerColor <=> this.styler.color;
```

**Reply by Zxynine** [reply]
That is fair, but I also think the operator you are proposing is asking too much and for someone newer it would just be confusing. I feel that having it being called a setter implies clearly that you are setting the value without the need of  a proper asignment in such simple cases. Just like for the getter you do not specify return. I feel like there needs to be consistency in that and we should either be able to omit "= value" in setters or should be required to include "return " in getters.

**Reply by HaloFour** [reply]
The current syntax is consistent with how expression-bodied members works throughout the language, including with lambdas.  This syntax isn't special to properties, and in no other case would there be an implicit assignment.  Nor is a setter required to make such an assignment.

```cs
int getter() => this.styler.color;
void setter(Color value) => this.styler.color = value;

Func<Color> getter = () => this.styler.color;
Action<Color> setter = value => this.styler.color = value;
```

**Comment by TahirAhmadov** [comment]
Perhaps you are looking for something like this: 
https://github.com/dotnet/csharplang/discussions/1007

**Reply by Zxynine** [reply]
Yes and no, I dont think it is as practical for them to implement  '<=>' and it would be confusing to some, I think just having slightly more implicit setters is all thats needed.

**Comment by theunrepentantgeek** [comment]
> I feel that having it being called a setter implies clearly that you are setting the value 

Yup. But there's a wide variety of places you _could_ be assigning that value. 

Introducing a special case of "if you happen to be using a member field from this instance for the body of the sender, we'll assume you want to write the value there" is _introducing_ inconsistency, not reducing it.

> Just like for the getter you do not specify return. 

If you're writing old style **get** members, you _do_ specify **return**:

``` csharp
public string FullName
{
    get { return name; }
    set { name = value; }
}
```

It's only when using expression bodied members that you can omit the **return** keyword - as a syntactic shorthand that's consistent with prior use of `=>` for lambdas. For these, the type of the expression is the deciding factor: if it's **void**, there's no return value; for any other type, the value is returned.

> ... there needs to be consistency in that and we should either be able to omit "= value" in setters ...

Doing this would make the language a lot _less_ consistent. Nowhere else in the language do we have an expression implying assignment.

> or should be required to include "return " in getters.

You can already include **return** in getters if you want to write them using the original syntax. 
Making this required for expression-bodied getters would undo much (maybe all) of the beauty of expression bodied getters, as well as breaking _literally_ billions of lines of existing C# code for a reason best described as trivial.

**Comment by MillKaDe** [comment, marked as answer]
> ...
> EDIT:
> 
> I realised I provided too simple of an example that can just be implemented using {get; set;}, Here is another that should showcase why i made this suggestion:
> 
> ```
> 		public colour StylerColour{
> 			get => this.styler.color;
> 			set => this.styler.color = value
> 		}
> ```
> ...

If you want a property, that redirects read/get and write/set access to somewhere else (instead of a compiler generated backing field), try this:
```c#
public ref colour StylerColour => ref this.styler.color;
```

Some [examples at SharpLab.io](https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwAICYCMBYAUKgZgAIMiBhAHgBUA+IgbzyKKqIDEBuJk4sAUwBmLIgAUiAXjr8hnboXJEAFK2ABKCXTYSiwDszwBfPHgzFGuZqkwA2EgBYiAZQAWAewDu1Osp3rJJTABOJQASACJ6YABPAAc+VyFlVQMQBmADMNUuCwDbVAcqPgBnYC8lFXVzZmZKWiIAY20oPndyMsU4QQBDCAAbNT0AekGiAEsoUeBRrt7RgC8+HVciToEe/qIANxmIPm5mFw8leoA6ESyiYaIAcz5gQf4uuCIBUb5e561N6aIYsFc4mBoqJ9g0ztpdJcRkU7oN3GBJotXu9Plsfn8AXwgVFREsiM1WttersdKDDq1FKdztlmAZqnJiFY8g4ALJdcZKSqgwolJR2dAXK7jYDc4rAJQEE6YOwCQUjAS9VxdEU5Zg88WKMIAI16XWcmSGIysAAZRbzFASiAA1LFFUauKBKTAwIjoF0EF12VRZbhGXAGIA==) :

```c#
using System;
public class C<T> {
  T F;
  public ref T P => ref F;
  public C (T t) => F = t;  
}

public static class Program {
  static void Show<T> (T t) => Console.WriteLine ($"{typeof (T)}: {t}");
  static void Test<T> (T t) {
    C<T> c = new C<T> (default); // initialize to default value
    Show (c.P); // get/read field F via property P
    c.P = t; // set/write field F via property P to new value t
    Show (c.P);
  }   
  public static void Main () {
    Test (42); // int
    Test (3.14f); // float
    Test ("blah"); // string
    Test (new Version (1, 2, 3, 4));
  }
}
```
.. which prints ..

```
System.Int32: 0
System.Int32: 42
System.Single: 0
System.Single: 3.14
System.String: 
System.String: blah
System.Version: 
System.Version: 1.2.3.4
```

**Reply by theunrepentantgeek** [reply]
That's really cool - and an application of **ref** that hadn't occurred to me. 

Unfortunately, this doesn't work with properties, only fields. 

This doesn't work:

``` csharp
public class C<T> {
  private T F { get; set; }
  public ref T P => ref F;  // error CS8156: An expression cannot be used in this context because it may not be passed or returned by reference
  public C (T t) => F = t;  
}
```

**Reply by MillKaDe** [reply]
But this works :
```c#
public class C<T> {
  // T F;
  T FP { get; set; }
  // public ref T P => ref F;
  public T P { get => FP; set => FP = value; }
  public C (T t) => F = t;  
}
```

So a future compiler version could avoid CS8156 by 'lowering' this
```c#
  public ref T P => ref FP; // error CS8156
```
or maybe even this
```c#
  public T P <=> FP;
```
to that
```c#
  public T P { get => FP; set => FP = value; }
```

Properties are just syntactic sugar to access something as if it is a field, even if that something is more complex than a field.
So its more a question of how much syntactic sugar is OK.

**Reply by svick** [reply]
@MillKaDe A `ref`-returning property is not the same as a read-write property. For example, only the result of a `ref-`returning property can be assigned to a `ref` local.

**Reply by TahirAhmadov** [reply]
Yes, `ref` keyword is not "syntactic sugar" for something else - it has a specific meaning of creating a new type. `ref int` means `typeof(int).MakeByRefType()`.

**Reply by sab39** [reply]
> or maybe even this
> 
> ```cs
>   public T P <=> FP;
> ```

Pretty sure that's been proposed before but I don't remember the issue or discussion to link to it.

**Reply by jnm2** [reply]
@sab39 See https://github.com/dotnet/csharplang/discussions/6018#discussioncomment-2570771
